buildbot/master/files/config/factories.py

```diff
@@ -103,6 +103,86 @@ def make_step(self, command):
         return step_class(**step_kwargs)
 
 
+class StepsYAMLParsingStep(buildstep.BuildStep):
+    """
+    Step which resolves the in-tree YAML and dynamically add test steps
+    """
+    yaml_path = ""
+
+    def __init__(self, builder_name, environment, yaml_path):
+        print(yaml_path)
+        self.builder_name = builder_name
+        self.environment = environment
+        self.yaml_path = yaml_path
+
+    def run(self):
+        print(self.yaml_path)
+        try:
+            full_yaml_path = os.path.join(self.build.workdir, self.yaml_path)
+            with open(full_yaml_path) as steps_file:
+                builder_steps = yaml.safe_load(steps_file)
+            commands = builder_steps[self.builder_name]
+            dynamic_steps = [self.make_step(command) for command in commands]
+        except Exception as e:  # Bad step configuration, fail build
+            print(str(e))
+            dynamic_steps = [BadConfigurationStep(e)]
+
+        # TODO: windows compatibility (use a custom script for this?)
+        pkill_step = [steps.ShellCommand(command=["pkill", "-x", "servo"],
+                                         decodeRC={0: SUCCESS, 1: SUCCESS})]
+
+        # util.BuildFactory is an old-style class so we cannot use super()
+        # but must hardcode the superclass here
+        self.build.steps = pkill_step + dynamic_steps
+
+    def make_step(self, command):
+        step_kwargs = {}
+        step_kwargs['env'] = self.environment
+
+        command = command.split(' ')
+        step_kwargs['command'] = command
+
+        step_class = steps.ShellCommand
+        args = iter(command)
+        for arg in args:
+            # Change Step class to capture warnings as needed
+            # (steps.Compile and steps.Test catch warnings)
+            if arg == './mach':
+                mach_arg = next(args)
+                if re.match('build(-.*)?', mach_arg):
+                    step_class = steps.Compile
+                elif re.match('test-.*', mach_arg):
+                    step_class = steps.Test
+
+            # Capture any logfiles
+            elif re.match('--log-.*', arg):
+                logfile = next(args)
+                if 'logfiles' not in step_kwargs:
+                    step_kwargs['logfiles'] = {}
+                step_kwargs['logfiles'][logfile] = logfile
+
+        return step_class(**step_kwargs)
+
+
+class DynamicServoIntreeYAMLFactory(ServoFactory):
+    """
+    Smart factory which takes a list of shell commands from a yaml file
+    and creates the appropriate Buildbot Steps. Uses heuristics to infer
+    Step type, if there are any logfiles, etc.
```

Comment:
Please update this comment to make it clear that the YAML file is coming from the main servo/servo repo (and capitalize YAML also).